server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/projection/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,156 @@
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.REMOTE_ADDR;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.REMOTE_HOST;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.SENDER;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.STATE;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.TABLE_NAME;
+import static org.apache.james.queue.rabbitmq.projection.cassandra.EnqueuedMailsTable.TIME_RANGE_START;
+
+import java.nio.ByteBuffer;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.mail.internet.AddressException;
+
+import org.apache.james.backends.cassandra.init.configuration.ClusterConfiguration;
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.core.MailAddress;
+import org.apache.james.queue.rabbitmq.projection.cassandra.model.CassandraMailQueueItem;
+import org.apache.james.util.CompletableFutureUtil;
+import org.apache.mailet.PerRecipientHeaders;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.UDTValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+public class EnqueuedMailsDao {
+
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement selectFrom;
+    private final CassandraUtils cassandraUtils;
+    private final Integer bucketSize;
+
+    @Inject
+    public EnqueuedMailsDao(Session session, CassandraUtils cassandraUtils, ClusterConfiguration clusterConfiguration) {
+        this.executor = new CassandraAsyncExecutor(session);
+        this.cassandraUtils = cassandraUtils;
+        this.bucketSize = clusterConfiguration.getHosts().size();
+
+        this.selectFrom = prepareSelectFrom(session);
+    }
+
+    private PreparedStatement prepareSelectFrom(Session session) {
+        return session.prepare(select()
+                .from(TABLE_NAME)
+                .where(eq(QUEUE_NAME, bindMarker(QUEUE_NAME)))
+                .and(gte(TIME_RANGE_START, bindMarker(TIME_RANGE_START)))
+                .and(eq(BUCKET_ID, bindMarker(BUCKET_ID))));
+    }
+
+    public CompletableFuture<Stream<CassandraMailQueueItem>> findEnqueuedMailsFrom(String queueName, Long timeRangeStartFrom) {
+        return CompletableFutureUtil.allOf(IntStream
+                .range(0, bucketSize)
+                .mapToObj(bucketId -> findEnqueuedMailsFrom(queueName, timeRangeStartFrom, bucketId)))
```

Comment:
It looks expensive

Reply:
Not really, you need to list queue content, and you spread the load upon writes between the different buckets. Thus you need to read them all.

Issue is the name of the method (which is actually `browse`)

Reply:
You need to filter out deleted mails using the specific projection table I described earlier.